src/deploy/functions/params.ts

```diff
@@ -168,8 +185,108 @@ export interface SelectInput<T> {
   select: Array<SelectOptions<T>>;
 }
 
-export type Param = StringParam | IntParam;
-type ParamValue = string | number | boolean;
+// Future supported resource types will be added to this literal type. Tooling SHOULD fall back
+// to text entry if it encounters an unknown ResourceParamType
+type ResourceType = "storage.googleapis.com/Bucket" | string;
+
+interface ResourceInput {
+  type: "resource";
+
+  resource: {
+    type: ResourceType;
+  };
+}
+
+interface SecretParam {
+  type: "secret";
+
+  // name of the param. Will be exposed as an environment variable with this name
+  name: string;
+
+  // A human friendly name for the param. Will be used in install/configure flows to describe
+  // what param is being updated. If omitted, UX will use the value of "param" instead.
+  label?: string;
+
+  // A long description of the parameter's purpose and allowed values. If omitted, UX will not
+  // provide a description of the parameter
+  description?: string;
+}
+
+export type Param = StringParam | IntParam | BooleanParam | SecretParam;
+type RawParamValue = string | number | boolean;
+
+/**
+ * A type which contains the resolved value of a param, and metadata ensuring
+ * that it's used in the correct way:
+ * - ParamValues coming from a dotenv file will have all three legal type fields set
+ *   but, we currently shallow copy the ParamValue and set the copy's type fields when
+ *   assigning a dotenv ParamValue to a specific Param.
+ * - ParamValues coming from prompting a param will have type fields corresponding to
+ *   the type of the Param.
+ * - ParamValues coming from Cloud Secrets Manager will have a string type field set
+ *   and isSecret = true, telling the Build process not to write the value to .env files.
+ */
+export class ParamValue {
+  rawValue: string;
+  isSecret: boolean;
+
+  constructor(
+    rawValue: string,
+    secret: boolean,
+    types: { string?: boolean; boolean?: boolean; number?: boolean }
+  ) {
+    this.rawValue = rawValue;
+    this.isSecret = secret;
+    this.legalString = types.string || false;
+    this.legalBoolean = types.boolean || false;
+    this.legalNumber = types.number || false;
+  }
+
+  // Call this when we figure out the actual type of a ParamValue from the .envs
+  narrowType(isType: Param["type"]): void {
```

Comment:
nit* `isType` as throwing me off b/c I expected it to be a boolean value per naming convention but actually is a string.

Reply:
hmm, well, `paramType` is presumably pretty inoffensive....